### PR TITLE
Launch and reconfig gas optimization 2

### PR DESCRIPTION
**Pull request by mejango**
pack distributionLimit and overflowAllowance values with their respective currencies

Screen Shot 2022-01-22 at 2 15 10 PM

### CODE REVIEW

File: contracts/JBController.sol [hunk 1]
```diff
@@ -136,48 +160,72 @@ contract JBController is IJBController, JBTerminalUtility, JBOperatable, Reentra
...
   */
-  mapping(uint256 => mapping(uint256 => mapping(IJBTerminal => uint256)))
-    public
-    override distributionLimitOf;
+  function distributionLimitCurrencyOf(
+    uint256 _projectId,
+    uint256 _configuration,
+    IJBTerminal _terminal
+  ) external view override returns (uint256) {
+    return _packedOverflowAllowanceDataOf[_projectId][_configuration][_terminal] >> 248;
```
Comment (dhyon) [inline contracts/JBController.sol:195]:
```suggestion
    return _packedDistributionLimitDataOf[_projectId][_configuration][_terminal] >> 248;
```

**Review by dhyon — COMMENTED** [review]
Dope - this is a fairly seamless change

File: contracts/JBController.sol (hunk 1 above)
Comment (mejango) [reply contracts/JBController.sol]: nice. fixed.

File: contracts/JBController.sol [hunk 2]
```diff
@@ -825,20 +873,36 @@ contract JBController is IJBController, JBTerminalUtility, JBOperatable, Reentra
...
-          _constraints.terminal
-        ] = _constraints.overflowAllowance;
+      if (_constraints.overflowAllowance > 0) {
+        // The currency of the overflow allowance should fit in a uint248.
+        require(_constraints.overflowAllowance < type(uint248).max, '0x00: BAD_OVERFLOW_ALLOWANCE');
+
+        // The currency of the overflow allowance should fit in a uint8.
+        require(
+          _constraints.overflowAllowanceCurrency < type(uint8).max,
+          'BAD_OVERFLOW_ALLOWANCE_CURRENCY'
```
Comment (dhyon) [inline contracts/JBController.sol:899]: We should use '0x00: BAD_OVERFLOW_ALLOWANCE_CURRENCY' until the custom errors from @viraj124 come in

**Review by dhyon — APPROVED** [review]

File: contracts/JBController.sol (hunk 2 above)
Comment (mejango) [reply contracts/JBController.sol]: nice. done.

File: contracts/JBController.sol [hunk 3]
```diff
@@ -107,6 +107,30 @@ contract JBController is IJBController, JBTerminalUtility, JBOperatable, Reentra
   */
   mapping(uint256 => int256) private _processedTokenTrackerOf;
 
+  /**
+    @notice 
+    bits 0-248: The amount of that a project can withdraw per funding cycle.
```
Comment (drgorillamd) [inline contracts/JBController.sol:112]: nit> 
```suggestion
    bits 0-248: The amount of token that a project can withdraw per funding cycle.
```

File: contracts/JBController.sol [hunk 4]
```diff
@@ -107,6 +107,30 @@ contract JBController is IJBController, JBTerminalUtility, JBOperatable, Reentra
...
+  mapping(uint256 => mapping(uint256 => mapping(IJBTerminal => uint256)))
+    public _packedDistributionLimitDataOf;
+
+  /**
+    @notice 
+    bits 0-247: The amount of overflow that a project is allowed to tap into on-demand throughout configuration.
+    bits 248-255: The currency of the amount of overflow that a project is allowed to tap.
+
+    _projectId The ID of the project to get the current overflow allowance of.
+    _configuration The configuration of the during which the allowance applies.
```
Comment (drgorillamd) [inline contracts/JBController.sol:128]: nit>
```suggestion
    _configuration The configuration during which the allowance applies.
```

File: contracts/JBController.sol [hunk 5]
```diff
@@ -136,48 +160,72 @@ contract JBController is IJBController, JBTerminalUtility, JBOperatable, Reentra
...
 
   //*********************************************************************//
-  // --------------------- public stored properties -------------------- //
+  // ------------------------- external views -------------------------- //
   //*********************************************************************//
 
   /**
     @notice 
-    The amount of overflow that a project is allowed to tap into on-demand throughout configuration.
+    The amount of that a project can withdraw per funding cycle.
```
Comment (drgorillamd) [inline contracts/JBController.sol:168]:
```suggestion
    The amount of token that a project can withdraw per funding cycle.
```

**Review by drgorillamd — APPROVED** [review]

**Review by dhyon — APPROVED** [review]

**Review by drgorillamd — APPROVED** [review]

**Pull request merged**